File: Pascal's_Triangle.swift
/*
Given a non-negative integer numRows, generate the first numRows of Pascal's triangle.


In Pascal's triangle, each number is the sum of the two numbers directly above it.

Example:

Input: 5
Output:
[
     [1],
    [1,1],
   [1,2,1],
  [1,3,3,1],
 [1,4,6,4,1]
]
*/

/*
Solution 1

keep pre to store previous array

Time Complexity: O(numRows^2)
Space Complexity: O(numRows)
*/
class Solution {
    func generate(_ numRows: Int) -> [[Int]] {
        guard numRows > 0 else { return [] }
        if numRows == 1 { return [[1]] }
        
        var res: [[Int]] = [[1]]
        var pre = [1]
        for i in 1..<numRows {
            let temp = pre
            for k in 0..<(pre.count-1) {
                pre[k+1] = temp[k] + temp[k+1]
            }
            pre.append(1)
            
            res.append(pre)
        }
        
        return res
    }
}

File: Pascal's_Triangle.cpp
/*
Given numRows, generate the first numRows of Pascal's triangle.

For example, given numRows = 5,
Return

[
     [1],
    [1,1],
   [1,2,1],
  [1,3,3,1],
 [1,4,6,4,1]
]
*/

class Solution {
 public:
  vector<vector<int>> generate(int numRows) {
    vector<vector<int>> ntriangle(numRows);

    for (int i = 0; i < numRows; i++) {
      ntriangle[i].resize(i + 1);  // resize the i th element in the ntriangle
      ntriangle[i][0] = 1;
      ntriangle[i][i] = 1;
      for (int j = 1; j < i; j++) {
        ntriangle[i][j] = ntriangle[i - 1][j - 1] + ntriangle[i - 1][j];
      }
    }
    return ntriangle;
  }
};
